File: 07-name-that-number/namenum.py
"""
ID: caleb.h1
LANG: PYTHON3
PROG: namenum
"""

def main():
    keypad = {
        '2': {'A','B','C'},
        '3': {'D','E','F'},
        '4': {'G','H','I'},
        '5': {'K','J','L'},
        '6': {'M','N','O'},
        '7': {'P','R','S'},
        '8': {'T','U','V'},
        '9': {'W','Y','Z'},
    }

    with open("dict.txt", 'r') as f:
        possibleNames = [i.strip() for i in f.readlines()]

    with open("namenum.in", 'r') as f:
        numbers = f.readline().strip()

    # The purge - Remove from the possibleNames ones that don't fit

    # Remove ones too long or too short
    i = 0
    while i < len(possibleNames):
        if len(possibleNames[i]) != len(numbers):
            possibleNames.pop(i)
        else:
            i+=1

    # Remove those that don't share letters
    for i in range(len(numbers)):
        letters = keypad[numbers[i]]
        j = 0
        while j < len(possibleNames):
            name = possibleNames[j]
            if name[i] not in letters:
                possibleNames.pop(j)
            else:
                j += 1


    # Write NONE if no names match, otherwise all matching names
    if not possibleNames:
        open("namenum.out", 'w').write("NONE\n")
    else:
        print(*possibleNames, sep='\n', file=open("namenum.out", 'w'))

if __name__ == '__main__':
    main()

File: 07-name-that-number/namenum.cpp
/*
ID: caleb.h1
LANG: C++
PROG: namenum
*/

#include <fstream>
#include <unordered_set>
#include <set>
#include <unordered_map>
#include <string>
#include <array>

#include <iostream>


int main() {
    static std::unordered_map<char, std::array<char, 3>> keypad = {
        {'2', {'A','B','C'}},
        {'3', {'D','E','F'}},
        {'4', {'G','H','I'}},
        {'5', {'K','J','L'}},
        {'6', {'M','N','O'}},
        {'7', {'P','R','S'}},
        {'8', {'T','U','V'}},
        {'9', {'W','Y','Z'}},
    };

    std::ifstream infile("dict.txt");
    std::set<std::string> possibleNames;
    std::string name;
    while (getline(infile, name)) possibleNames.insert(name);

    infile.close();
    infile.open("namenum.in");

    char numbers[15];
    char numbersCount = 0;
    char c;
    while (infile >> c) {
        numbers[numbersCount++] = c;
    }

    infile.close();

    // The purge - Remove from the possibleNames ones that don't fit

    // Remove ones too long or too short
    for (auto it = possibleNames.begin(); it != possibleNames.end();)
        if (it->length() != numbersCount)
            it = possibleNames.erase(it);
        else
            ++it;

    // Remove those that don't share letters
    for (char i = 0; i < numbersCount; ++i) {
        std::array<char, 3> a = keypad[numbers[i]];
        for (auto it = possibleNames.begin(); it != possibleNames.end();) {
            if ((*it)[i] != a[0] && (*it)[i] != a[1] && (*it)[i] != a[2])
                it = possibleNames.erase(it);
            else
                ++it;
        }
    }


    // Write NONE if no names match, otherwise all matching names
    std::ofstream outfile("namenum.out");

    if (!possibleNames.size())
        outfile << "NONE" << std::endl;
    else
        for (std::string name : possibleNames)
            outfile << name << std::endl;
    
    outfile.close();

    return 0;

}
